jobserver/models/reports.py

```diff
@@ -92,3 +94,138 @@ def __str__(self):
+                            approved_by__isnull=False,
+                        )
+                    )
+                ),
+                name="%(app_label)s_%(class)s_both_approved_at_and_approved_by_set",
+            ),
+            models.CheckConstraint(
+                check=(
+                    Q(
+                        created_at__isnull=True,
+                        created_by__isnull=True,
+                    )
+                    | (
+                        Q(
+                            created_at__isnull=False,
+                            created_by__isnull=False,
+                        )
+                    )
+                ),
+                name="%(app_label)s_%(class)s_both_created_at_and_created_by_set",
+            ),
+            models.CheckConstraint(
+                check=(
+                    Q(
+                        updated_at__isnull=True,
+                        updated_by__isnull=True,
+                    )
+                    | (
+                        Q(
+                            updated_at__isnull=False,
+                            updated_by__isnull=False,
+                        )
+                    )
+                ),
+                name="%(app_label)s_%(class)s_both_updated_at_and_updated_by_set",
+            ),
+        ]
+
+    @classmethod
+    def create_from_report(cls, *, report, user):
+        if report is None:
+            return None
+
+        if not hasattr(report, "analysis_request"):
+            raise Exception("bad type of report??")
+
+        with transaction.atomic():
+            # create a request to publish the released file underpinning the report
+            rfile_publish_request = ReleaseFilePublishRequest.objects.create(
+                created_by=user, workspace=report.release_file.workspace
+            )
+            rfile_publish_request.files.add(report.release_file)
+
+            # create a request to publish the report
+            return ReportPublishRequest.objects.create(
+                report=report,
+                release_file_publish_request=rfile_publish_request,
+                created_by=user,
+                updated_by=user,
+            )
+
+    def __str__(self):
+        return f"Publish request for report: {self.report.title}"
+
+    @transaction.atomic()
+    def approve(self, *, user):
+        now = timezone.now()
+
+        self.approved_at = now
+        self.approved_by = user
+        self.save(update_fields=["approved_at", "approved_by"])
+
+        self.report.published_at = now
+        self.report.published_by = user
+        self.report.save(update_fields=["published_at", "published_by"])
+
+        self.release_file_publish_request.approve(user=user, now=now)
+
+    def get_absolute_url(self):
+        return reverse(
+            "interactive:report-publish-request-update",
+            kwargs={
+                "org_slug": self.report.analysis_request.project.org.slug,
```

Comment:
It feels weird that we're having to ask the AnalysisRequest for the org and project details. Further up we're also getting the workspace from `report.release_file.workspace`, which feels a little better but is inconsistent if we're sometimes asking ReleaseFile and sometimes AnalysisRequest.

I feel that the ReportPublishRequest should be asking the Report for the org and project at least. It seems like a Report should belong to a Project, I think?!

Reply:
That's a really good point!  That tracks nicely for both type of report too, because even if we don't show the project on a curated report, it should still be under the umbrella of a report under the hood so that a) it can be listed for a Project easily, and b) because the Project represents the the IG cover.